torch/distributed/fsdp/_init_utils.py

```diff
@@ -746,14 +746,34 @@ def _get_buffer_names(root_module: nn.Module) -> Set[str]:
 def _check_single_device_module(
     module: nn.Module,
     ignored_params: Set[nn.Parameter],
+    device_id: Optional[Union[int, torch.device]],
 ) -> None:
     """
     Raises an error if ``module`` has original parameters on multiple devices,
     ignoring the parameters in ``ignored_params``. Thus, after this method, the
     module must be either fully on the CPU or fully on a non-CPU device.
     """
     devices = {param.device for param in _get_orig_params(module, ignored_params)}
-    if len(devices) > 1:
+    # We allow module to be partially on CPU and partially on GPU if device_id is not
+    # None, since the device_id arg will result in the CPU portion being moved to
+    # GPU. This is useful in cases where part of the module may be parallelized
+    # by another algorithm and may already be on GPU. We'd like to enforce device_id
+    # to not be None, otherwise we'd flatten parameters in a mixed module which is
+    # not supported.
+    if (
+        len(devices) == 2 and torch.device("cpu") in devices
+    ):
+        if device_id is not None:
+            warnings.warn(
+                f"FSDP wrapping module with params on multiple devices: {devices}"
+                "This could occur in cases where another parallelism algorithm has already"
+                "moved some parameters to a non-CPU device."
```

Comment:
I am not sure if we should give this warning.

What I am thinking is that one of the following cases holds:
- The user knows that he/she is using mixture of parallelisms, and then this warning is not too informative.
- The user does not know that he/she is using a mixture of parallelisms, and then this warning may be jarring.
- the user is not using a mixture of parallelisms but hits this branch somehow, and then this warning may be confusing.